pkg/operator/configobservation/etcd/observe_etcd.go

```diff
@@ -23,6 +23,14 @@ const (
 // then the previously observed storage config URLs will be re-used.
 func ObserveStorageURLs(genericListers configobserver.Listers, recorder events.Recorder, currentConfig map[string]interface{}) (map[string]interface{}, []error) {
 	listers := genericListers.(configobservation.Listers)
+
+	// If the cluster has already been migrated to the endpoints configmap, no-op
+	if _, err := listers.ConfigmapLister.ConfigMaps(etcdEndpointNamespace).Get("etcd-endpoints"); err == nil {
```

Comment:
This is racy and risks breakage if a trigger happens for a different reason and one observer sees the configmap and the other doesn't

Reply:
Do you have some suggestion for how to implement this safely?

Reply:
I'm not sure I understand the impact of the race. If the endpoints version doesn't see the configmap and continues to try and read the endpoint, the controller will:

1. Fail to find the endpoint because this is a new cluster where the endpoint never existed and will never exist, and on retry the cm should eventually be observed and this func should begin correctly no-op'ing
2. Fail to find the endpoint because this is an upgrade and the cm either hasn't been created yet or again the cache is stale (in which case either read the old endpoint which is still valid, and eventually start no-op'ing once the cm is available)

Help me understand the race impact?

Reply:
@ironcladlou if the endpoint is found and the configmap is found, then two configobservers will try to set the same fields and we go degraded on that condition.  If it doesn't happen for very long, this may be acceptable.

Reply:
Two ways I can think of to address this.

The original code I had (https://github.com/openshift/cluster-kube-apiserver-operator/pull/859/commits/582e221c1368bb5536bb08b2e19f981753247a50) registers only the configmap observer, which itself delegates to the endpoints observer if the configmap isn't found.

Another possible way is to refactor the endpoints observer code to internally handle the existence of either the endpoints or configmap. I have a commit started for this but it's pretty invasive to the current endpoints handling code.

If https://github.com/openshift/cluster-kube-apiserver-operator/pull/859/commits/582e221c1368bb5536bb08b2e19f981753247a50 is okay I can revert to that.

Reply:
Alright, I refactored to an approach which uses a delegation model to ensure only one implementation is ever used at a time depending on whether the configmap exists during the observation attempt